### PR TITLE
Fixes #1651 - BP: Create batch queue management interface

### LINKED ISSUE #1651: BP: Create batch queue management interface
We need a way for the GIS team to manage the queue internally.

- [x] Create a "Manage Queue" tab in the BP modal that will only be accessible if the client is hosted on NatWeb (https://staging-apps.usgs.gov/streamstats/bp)
- [x] Use this GET request: https://streamstats.usgs.gov/batchprocessor/batch/?excludeDeleted=true to retrieve all batches that aren't soft-deleted
- [x] Create a table on the "Manage Queue" tab:
     - [x] Include all columns in the table on the "Batch Status" column
     - [x] Add an "Order" column to the right of the Batch ID column. This should be an editable text field that is populated with the Order property of the batch
     - [x] The "Status" column should show the "Description" instead of the "Message"
          - [x] Service work needed: https://code.usgs.gov/StreamStats/web-services-and-apis/ss-batchprocessor/-/issues/74
     - [x] Add an "Email Address" field to the right of the Order column. This should contain the email address associated with the batch.
     - [x] Add a "Pause?" button to the left of the "Delete?" column. If the batch is paused (Status ID = 9), show the play button (`<i class="fa fa-play-circle-o"></i>`). If the batch is in the queue (Status ID = 1), show the pause button icon `<i class="fa fa-pause-circle-o"></i>`. If the batch has any other status, show nothing.
 - [x] Add an "Update Queue Order" Button at the bottom of the tab
 - [x] When the user clicks the "Update Queue Order" Button, make a POST request to https://streamstats.usgs.gov/batchprocessor/batch/order with the current Batch IDs and Orders of each batch (each row in the table)
      - POST body format:
```
{
  "batchOrder": [
    {
      "batchID": 200,
      "order": 3
    },
    {
      "batchID": 201,
      "order": 2
    },
    {
      "batchID": 202,
      "order": 1
    }
  ]
}
``` 
- [x] When a user clicks the Pause button, make a POST request to https://streamstats.usgs.gov/batchprocessor/pauseBatch?batchID={batchID}
- [x] When a user clicks the Play button, make a POST request to https://streamstats.usgs.gov/batchprocessor/unpauseBatch?batchID={batchID}
     - [x] Service work needed: https://code.usgs.gov/StreamStats/web-services-and-apis/ss-batchprocessor/-/issues/73

### PR DESCRIPTION
**Pull request by amedenblik**
Closes #1651

https://code.usgs.gov/StreamStats/web-services-and-apis/ss-batchprocessor/-/merge_requests/44 should be reviewed, merged, and deployed first before review of this PR  (Now done! ✅)

### CODE REVIEW

**Comment by amedenblik** [comment]
Deployment instructions are [here in OneNote](https://doimspp.sharepoint.com/:o:/r/sites/GS-UMidWIM/_layouts/15/Doc.aspx?sourcedoc=%7B86ee3545-8271-46a1-b16f-158b12f19ba3%7D&action=edit&wd=target(FAQs.one%7CD8FCAEF7-3539-4FED-A0C2-5BCBE0E4F155%2FHow%20do%20I%20deploy%20the%20StreamStats%20Batch%20Processor%20Queue%20Management%7CE291213F-F375-4740-A836-F56330A84F73%2F)onenote%3Ahttps%3A%2F%2Fdoimspp.sharepoint.com%2Fsites%2FGS-UMidWIM%2FShared%20Documents%2FProjects%2FStreamstats%20Ecosystem%2FKJ%27s%20Guide%20to%20StreamStats%201%2FFAQs.one#How%20do%20I%20deploy%20the%20StreamStats%20Batch%20Processor%20Queue%20Management%26section-id%3D%7BD8FCAEF7-3539-4FED-A0C2-5BCBE0E4F155%7D%26page-id%3D%7BE291213F-F375-4740-A836-F56330A84F73%7D%26end).

**Comment by amedenblik** [comment]
I deployed to https://apps-int.usgs.gov/streamstats/bp/ temporarily to confirm the new tab appears on NatWeb.

**Comment by amedenblik** [comment]
Deployed to https://staging-apps.usgs.gov/streamstats/bp for testing and the "Manage Batch" tab does appear, as expected.

Also temporarily deployed to https://dev.streamstats.usgs.gov/bp-temp/ for testing and the "Manage Batch" tab doesn't appear, as expected. I'll delete this site once the PR is merged.

File: src/Views/batchprocessor.html [hunk 1]
```diff
@@ -255,62 +273,175 @@ <h3>Select Basin Characteristics</h3>
...
+        </div>
+        <!-- End batch status -->
+        <!-- Start manage queue -->
+        <div id="manageQueue" ng-if="vm.selectedBatchProcessorTabName==='manageQueue'">
+            <!-- Start instructions -->
+            <div>
+                <b>How to Edit the Batch Queue</b>
+                <ul>
+                    <li>To change the order of batches, edit the values in the <i>Order</i> column and click the "Update
+                        Queue Order" button. Batches will proceed in the queue in numerical order.</li>
```
Comment (harper-wavra) [inline src/Views/batchprocessor.html]: "Update Queue Order" button should this be 'Reorder Queue" button

File: src/Views/batchprocessor.html [hunk 2]
```diff
@@ -255,62 +273,175 @@ <h3>Select Basin Characteristics</h3>
...
+                                <th style="text-align: center;"><small>Number Points Successful</small></th>
+                                <th style="text-align: center;"><small>Pause / Unpause?</small></th>
+                                <th style="text-align: center;"><small>Delete?</small></th>
+                            </tr>
+                        </thead>
+                        <tbody ng-repeat="batch in vm.manageQueueList">
+                            <tr>
+                                <td style="text-align: center;"><span>{{batch.batchID}}</span></td>
+                                <td style="text-align: center;"><span>
+                                    <input ng-if="batch.status==1" class="bp-input" type="number" ng-model="batch.order" style="text-align: center; width:50px;" required></span>
```
Comment (harper-wavra) [inline src/Views/batchprocessor.html]: I wonder if we should limit this to positive numbers.

**Review by harper-wavra — CHANGES_REQUESTED** [review]
I think the functionality is all there which is great! 

- I think we might need to make the order column a bit bigger. If we get numbers past two digits I am not able to see all the numbers. [example](https://github.com/USGS-WiM/StreamStats/assets/59063181/724623dc-ee08-42fe-8954-b6653f71c1b3)

- I am not sure if we need time completed, results, number of successful points in this table since it only shows batches that haven't been run. If we remove these it might give us more room for the order column. 

- When I reorder a queue and it fails is it possible to revert the number in the order column to the original order?

- This is more a service question, should I be able to change an order a number beyond the order, for example 10000? I think Kitty might be used to doing this to pause a batch.

- Instead of sorting by batch ID, does it make more since to have the batches sorted in ascending order? For example, order 1 on top, then 2, ect, then the paused batches on the bottom. Not sure if that is a ton more work & could just be an enhancement.

**Comment by amedenblik** [comment]
> * I think we might need to make the order column a bit bigger. If we get numbers past two digits I am not able to see all the numbers. [example](https://github.com/USGS-WiM/StreamStats/assets/59063181/724623dc-ee08-42fe-8954-b6653f71c1b3)

Done! ✅

> * I am not sure if we need time completed, results, number of successful points in this table since it only shows batches that haven't been run. If we remove these it might give us more room for the order column.

This list is supposed to show all batches that aren't in the Delete status. We just don't have any completed batches in the database at the moment. I've asked Kitty if it would be helpful to see the completed batches, and I am guessing it would be helpful so she can see how things have progressed.

> * When I reorder a queue and it fails is it possible to revert the number in the order column to the original order?

Yes, this is easy to do. But what if someone carefully reorders everything and then realizes they made a small mistake (ex. accidentally gave 2 batches the same Order number) after it fails? Then they have to redo everything. 

> * This is more a service question, should I be able to change an order a number beyond the order, for example 10000? I think Kitty might be used to doing this to pause a batch.

I don't think we've written this part of the services yet; it will be part of https://code.usgs.gov/StreamStats/web-services-and-apis/ss-batchprocessor/-/issues/8. When a worker is available, I think Celery is going to have a function that determines which batch should be run next. I was thinking we would just select whatever batch had the lowest Order number. Let me know if you have thoughts about that. 

> * Instead of sorting by batch ID, does it make more since to have the batches sorted in ascending order? For example, order 1 on top, then 2, ect, then the paused batches on the bottom. Not sure if that is a ton more work & could just be an enhancement.

Done! ✅


Redeployed to:
- https://staging-apps.usgs.gov/streamstats/bp/
- https://dev.streamstats.usgs.gov/bp-temp

**Comment by harper-wavra** [comment]
https://staging-apps.usgs.gov/streamstats/bp/ doesnt seem to be working for me, I seeing this [console error](https://github.com/USGS-WiM/StreamStats/assets/59063181/112cac30-978b-44b0-a74b-8840873b68e7). 

> When I reorder a queue and it fails is it possible to revert the number in the order column to the original order?

I wonder if instead of having the Order as an input box right away, they are first just labels. Then there is a button to reorder which would turn them into input boxes. Then the user can reorder and click submit. If it works, they turn back into labels & if it doesn't they stay as input boxes with the users entered changes. 
It is also a little confusing when the rows changes right when I edit the order number. If we go with the above change maybe we could leave the rows where they are until the user has entered submit and it was successful.

**Comment by harper-wavra** [comment]
Not sure if you fixed it or not, but I can get to https://staging-apps.usgs.gov/streamstats/bp/ now

**Comment by amedenblik** [comment]
> Not sure if you fixed it or not, but I can get to https://staging-apps.usgs.gov/streamstats/bp/ now

I didn't change anything! Interesting

**Comment by amedenblik** [comment]
> I wonder if instead of having the Order as an input box right away, they are first just labels. Then there is a button to reorder which would turn them into input boxes. Then the user can reorder and click submit. If it works, they turn back into labels & if it doesn't they stay as input boxes with the users entered changes.

I made this change

> It is also a little confusing when the rows changes right when I edit the order number. If we go with the above change maybe we could leave the rows where they are until the user has entered submit and it was successful.

I could go either way on this. It is a little confusing, but it's also nice because then you can also visualize the new order from top to bottom. Maybe we seek out a third opinion?

**Review by harper-wavra — COMMENTED** [review]
I am good with letting the GIS team feel out the moving rows, it does make it nice to see the order before it gets submitted.

The only thing I think is missing is a 'Cancel' button when I am recording.  If I click Reorder Queue, then decide I don't want to anymore the only way to get back is to close the modal and reopen it. I think when user clicks reorder queue, the submit button should appear like it does but maybe we also need a cancel button to appear as well.

**Comment by amedenblik** [comment]
Added cancel button and pulled in latest from `batch-processor` branch: [see changes here](https://github.com/USGS-WiM/StreamStats/compare/5411bcb7fc4ea97b2e02ffd4fa7abdf7c903a275...c12bc13a90c3e1dcb82d77a21d35fa28f449cb39)

**Review by harper-wavra — APPROVED** [review]

**Comment by amedenblik** [comment]
Deployed to:
- https://dev.streamstats.usgs.gov/bp/
- https://staging-apps.usgs.gov/streamstats/bp/

**Pull request merged**